src/hardware/gus.cpp

```diff
@@ -600,6 +601,9 @@ Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
 		voices.at(i) = std::make_unique<Voice>(i, voice_irq);
 	}
 
+	// Size the accumulating write buffer
+	accumulator.resize(BUFFER_SAMPLES);
```

Comment:
I am still as confused about this `accumulator` name as I was since the very beginning. Accumulator is a single value, not a collection of values.

Aside of that, this resize seems **very suspicious**. Shouldn't it be a `reserve`? 

TODO for me to check later in review: what is this accumulator anyway? Does reallocating some vector here break other code?

Reply:
Good point; it's an array (or now vector) of accumulators.  I'll try to rename it to be clearer.

If you look at the callback, it asks for ~1ms of audio.

When we enter the callback, we rendering one voice at a time:

``` c++
	auto v = voices.begin();
	const auto v_end = v + active_voices;
	while (v < v_end && *v) {
		v->get()->GenerateSamples(accumulator, ram, vol_scalars,
		                          pan_scalars, frames, dac_enabled);
		++v;
	}
```

`GenerateSamples` renders the audio for a single voice - say, a trumpet, over the requested 1ms worth of audio frames.

Notice we pass the `accumulator` array into `GenerateSamples(..)`. 

If we drill into GenerateSamples, you'll see the accumulator is now called 'stream' (because relative to this function, it's no longer an accumulator).  The stream values are `v` (unfortunately named for this example..) 

 But anyway, we see the trumpet's samples steadily are added to the accumulator values (the `*v ++ =`). It's not just an assignment.

``` c++
	// Add the samples to the stream, angled in L-R space
	while (v < last_v) {
		float sample = GetSample(ram);
		sample *= PopVolScalar(vol_scalars);
		if (dac_enabled) {
			*v++ += sample * pan_scalar.left;
			*v++ += sample * pan_scalar.right;
		}
	}
```

So now we've got accumulator holding 1 ms worth of trumpet.

We then step into next voice, say a drum, and literally add those samples onto the prior trumpet samples. 

That repeats for all active voices (up to 32 voices) until each accumulator [] position in the array has the sum of every voice :-)

Reply:
Regarding renaming a couple of these poorly named variables, I placed this in a separate commit: 4c040f7c5

Reply:
> Resize seems very suspicious. Shouldn't it be a reserve?

We only need to allocate the rendering buffer once (So I prefer to do that here in the GUS's constructor - one and for all).  After that, we reuse that array over and over, and never need to grow or shrink it.  

In general, for audio code you never want to do dymanic memory operations.  Great article here - it's all common sense, but it's nice and quick reading from someone who writes audio code for live concerts (one glitch, and your code will never be used again!). - http://atastypixel.com/blog/four-common-mistakes-in-audio-development/

`Reserve` does allocate the memory, but it doesn't let me use the elements until I  `push_back(elem)` (which I'm not going to do) or `.resize(elems)`.  So I'd rather just resize() straighy away.

Also, I resize by a `constexpr` value - so this allocation might even be optimized at compile-time.

Reply:
> Aside of that, this resize seems very suspicious. Shouldn't it be a reserve?

Yes - much better now:

``` c++
Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
        : render_buffer(BUFFER_FRAMES * 2), // 2 samples/frame (left & right)
```

And gets rid of the need for BUFFER_SAMPLES now!